### PR TITLE
support complex types for tanh_backward_cpu

### LINKED ISSUE #37701: Implement tanh_backward for complex dtypes
Implement tanh_backward for complex dtypes on cpu and cuda
```
>>> x = torch.randn(4, dtype=torch.complex64, requires_grad=True)
>>> y=torch.tanh(x)
>>> z=y.sum()
>>> z.backward()
[W python_engine.cpp:148] Warning: Complex backward is not fully supported yet and could lead to wrong gradients for functions we have not fixed yet (function operator())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chourdiaanjali/pytorch2/torch/tensor.py", line 184, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/chourdiaanjali/pytorch2/torch/autograd/__init__.py", line 115, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: "tanh_backward_cpu" not implemented for 'ComplexFloat' (operator() at aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.AVX2.cpp:517)
```

cc @ezyang @anjali411 @dylanbespalko

### PR DESCRIPTION
**Pull request by kshitij12345**
Closes: #37701 

TO-DO:
* [x] Add Tests

### CODE REVIEW

**Comment by kshitij12345** [comment]
@anjali411 
Please review.
Also is there another operator which has complex dtype test enabled in `test_autograd.py` for reference.

**Comment by anjali411** [comment]
@kshitij12345 this looks good! I am working on adding a whitelist for testing backwards for some C->C functions (will have a PR soon). so once that's pushed, you would be easily able to enable the test for complex dtype by adding tanh to the whitelist. 

should have the PR up by tomorrow

**Comment by kshitij12345** [comment]
@anjali411 Sounds good.

**Comment by anjali411** [comment]
hey @kshitij12345 you can add tanh to [this](https://github.com/pytorch/pytorch/blob/9232356e5fd7eb811c7980a88a93c7db578fce35/test/test_autograd.py#L4189) whitelist and then [this](https://github.com/pytorch/pytorch/blob/9232356e5fd7eb811c7980a88a93c7db578fce35/torch/testing/_internal/common_methods_invocations.py#L241) test should run with complex as well.

**Comment by kshitij12345** [comment]
@anjali411 
Is there any simple way to disable it for case when `device is cuda and dtype is complex` as `tanh` forward itself doesn't support `cuda` device.

**Comment by anjali411** [comment]
> @anjali411
> Is there any simple way to disable it for case when `device is cuda and dtype is complex` as `tanh` forward itself doesn't support `cuda` device.
> 
> Or should we add another `complex_cpu` list and update the test generation logic?
> 
> Would take care of `tanh_cuda` and `tanh_backward_cuda` in another PR.

Generally we could add "Expectedcudafailure" (used [here](https://github.com/pytorch/pytorch/blob/9232356e5fd7eb811c7980a88a93c7db578fce35/torch/testing/_internal/common_methods_invocations.py#L352)) if it were not working for floating and complex dtypes. And we might need to add something for that later, but for now I think we can just add a new test for complex `tanh_complex`,  add it to the whitelist and add "Expectedcudafailure" for it. And then that test would fail when we add tanh_backward for cuda so it would remind us to update that test

File: test/test_autograd.py [hunk 1]
```diff
@@ -4213,7 +4213,9 @@ def add_test(
             test_name = basic_test_name + ''.join('_neg' + str(i) for i, idx in enumerate(dim_perm) if idx < 0)
             if dtype.is_complex:
                 # TODO: remove this. this is temporary while we ramp up the complex support.
-                if name in complex_list and 'scalar' not in test_name and 'constant' not in test_name:
+                if 'complex' in test_name:
```
Comment (kshitij12345) [inline test/test_autograd.py:4216]: Using just `name` led to accepting all the `tanh` tests and not all of them fail on CUDA i.e. the one with standard `dtype double`.

Have instead added condition to check to filter based on the `test_name`.

Comment (anjali411) [reply test/test_autograd.py:4216]: yup makes perfect sense. I did something similar in this [PR](https://github.com/pytorch/pytorch/pull/38320) as well

**Comment by anjali411** [comment]
Also, would you like to work on adding cuda support for complex for a few other ops as well?

File: test/test_autograd.py (hunk 1 above)
Comment (kshitij12345) [reply test/test_autograd.py:4216]: Should we wait till #38320 is in so that we can use that machinery here. Otherwise I presume, there would be conflict.

**Comment by kshitij12345** [comment]
> Also, would you like to work on adding cuda support for complex for a few other ops as well?

Sure would. :)

File: test/test_autograd.py (hunk 1 above)
Comment (anjali411) [reply test/test_autograd.py:4216]: merged

**Comment by anjali411** [comment]
> > Also, would you like to work on adding cuda support for complex for a few other ops as well?
> 
> Sure would. :)

great! let's follow up on this on slack

File: test/test_autograd.py (hunk 1 above)
Comment (kshitij12345) [reply test/test_autograd.py]: Thanks. 
Done updating.

**Comment by kshitij12345** [comment]
@anjali411 PTAL :)

> great! let's follow up on this on slack
Sadly I haven't received the slack invite.

**Comment by kshitij12345** [comment]
@anjali411 Gentle ping.

**Review by anjali411 — APPROVED** [review]

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 2]
```diff
@@ -522,16 +522,31 @@ void sigmoid_backward_kernel(TensorIterator& iter) {
...
-    cpu_kernel_vec(iter,
-      [=](scalar_t a, scalar_t b) -> scalar_t {
-        return a * (scalar_t(1) - b * b);
-      },
-      [=](Vec256<scalar_t> a, Vec256<scalar_t> b) {
-        return a * (one_vec - b * b);
-      });
-  });
+  if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "tanh_backward_cpu", [&]() {
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:526]: we could combine this in future when `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES` starts using c10::complex
cc. @zasdfgbnm

**Pull request closed without merging**